File: LeetCodeKotlin/WordSearch_79.kt
class WordSearch_79 {
    fun exist(board: Array<CharArray>, word: String): Boolean {
        val rows = board.size
        val cols = board[0].size

        // Pre-check: Ensure the board contains enough instances of each character needed to form the word.
        val boardCount = hashMapOf<Char, Int>()
        for (row in board) {
            for (char in row) {
                boardCount[char] = boardCount.getOrDefault(char, 0) + 1
            }
        }
        for (char in word) {
            boardCount[char] = boardCount.getOrDefault(char, 0) - 1
            if (boardCount[char]!! < 0) {
                return false
            }
        }
        
        for (i in 0 until rows) {
            for (j in 0 until cols) {
                // Start DFS if first letter matches
                if (dfs(board, word, i, j, 0)) {
                    return true
                }
            }
        }
        return false
    }
    
    private fun dfs(board: Array<CharArray>, word: String, i: Int, j: Int, index: Int): Boolean {
        // If we've matched all characters, we've found the word
        if (index == word.length) return true
        
        // Check boundaries and if the current cell matches the word character at index
        if (i < 0 || i >= board.size || j < 0 || j >= board[0].size || board[i][j] != word[index]) {
            return false
        }
        
        // Temporarily mark the current cell as visited so we don't use it twice
        val temp = board[i][j]
        board[i][j] = '#'  // '#' is used as a placeholder
        
        // Explore in all four directions
        val found = dfs(board, word, i + 1, j, index + 1) ||
                    dfs(board, word, i - 1, j, index + 1) ||
                    dfs(board, word, i, j + 1, index + 1) ||
                    dfs(board, word, i, j - 1, index + 1)
        
        // Restore the cell's original character after the exploration
        board[i][j] = temp
        
        return found
    }
}

File: LeetCodeCpp/problems/backtracking/WordSearch_79.cpp
#include "WordSearch_79.h"

bool WordSearch_79::exist(vector<vector<char>>& board, string word)
{
    if (word.empty()) {
        return true;
    }

    if (board.empty() || board.front().empty()) {
        return false;
    }

    for (size_t row = 0; row < board.size(); ++row) {
        for (size_t col = 0; col < board[row].size(); ++col) {
            if (backtrack(board, word, row, col, 0)) {
                return true;
            }
        }
    }

    return false;
}

bool WordSearch_79::backtrack(vector<vector<char>>& board,
    const string& word,
    size_t row,
    size_t col,
    size_t index)
{
    if (board[row][col] != word[index]) {
        return false;
    }

    if (index + 1 == word.size()) {
        return true;
    }

    const char original = board[row][col];
    board[row][col] = '#';

    const size_t nextIndex = index + 1;

    const bool found =
        (row > 0 &&
         backtrack(board, word, row - 1, col, nextIndex)) ||
        (row + 1 < board.size() &&
         backtrack(board, word, row + 1, col, nextIndex)) ||
        (col > 0 &&
         backtrack(board, word, row, col - 1, nextIndex)) ||
        (col + 1 < board[row].size() &&
         backtrack(board, word, row, col + 1, nextIndex));

    board[row][col] = original;

    return found;
}
